File: quicksort.py
# QuickSort - Introduction to Algorithm version

def partition(arr, low, high):
	# select the last element of slice as pivot
	pivot = arr[high]
	i = low-1
	for j in range(low, high):
		if arr[j] <= pivot:
			i += 1
			arr[i], arr[j] = arr[j], arr[i]
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i+1

def quickSort(arr, low, high):
	if(low < high):
		mid = partition(arr, low, high)
		quickSort(arr, low, mid-1)
		quickSort(arr, mid+1, high)

def qSort(arr):
	toSort = arr
	quickSort(toSort, 0, len(toSort)-1)
	return toSort

# -------------------------------------- #

# Unit test
if __name__ == "__main__":
	# case 1
	case_1 = [9,8,7,6,5,4,3,2,1,0]
	sorted = qSort(case_1)
	case_1.sort()
	assert(cmp(case_1, sorted)==0)

	# case 2
	case_2 = [5,2,5,7,4,5,9,5,2,5,7]
	sorted = qSort(case_2)
	case_2.sort()
	assert(cmp(case_2, sorted)==0)

File: quicksort.go
package main

import (
	"fmt"
)

type slice []int

func partition(sls slice, low int, high int) int {
	pivot := sls[high]
	i := low-1
	for j:=low; j<high; j++ {
		if(sls[j] <= pivot){
			i++
			sls[i], sls[j] = sls[j], sls[i]
		}
	}
	sls[i+1], sls[high] = sls[high], sls[i+1]

	return i+1
}

func quickSort(sls slice, low int, high int){
	if(low < high){
		mid := partition(sls[:], low, high)
		quickSort(sls[:], low, mid-1)
		quickSort(sls[:], mid+1, high)
	}
}

func (arr slice) qSort(){
	quickSort(arr[:], 0, len(arr)-1)
}

func main(){
	// test case 1
	case1 := slice{9,8,7,6,5,4,3,2,1,0}
	case1.qSort()
	fmt.Println(case1)

	// test case 2
	case2 := slice{5,2,5,7,4,5,9,5,2,5,7}
	case2.qSort()
	fmt.Println(case2)
}